cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ApplicationLifecycleService.java

```diff
@@ -594,13 +592,52 @@ public ApplicationId upgradeApplication(ApplicationId appId,
       throw new ApplicationNotFoundException(appId);
     }
 
-    ApplicationId currentAppId = appId.getNamespaceId().app(currentSpec.getName(), currentSpec.getAppVersion());
+    return updateApplicationByArtifact(appId, currentSpec, allowedArtifactScopes, allowSnapshot);
+  }
+
+  /**
+   * Upgrades the latest version of an existing application
+   * by upgrading application artifact versions and plugin artifact versions.
+   *
+   * @param appRef the application reference of the application to upgrade.
+   * @param allowedArtifactScopes artifact scopes allowed while looking for latest artifacts for upgrade.
+   * @param allowSnapshot whether to consider snapshot version of artifacts or not for upgrade.
+   * @return the {@link ApplicationId} of the application version being updated.
+   * @throws IllegalStateException if something unexpected happened during upgrade.
+   * @throws IOException if there was an IO error during initializing application class from artifact.
+   * @throws JsonIOException if there was an error in serializing or deserializing app config.
+   * @throws UnsupportedOperationException if application does not support upgrade operation.
+   * @throws InvalidArtifactException if candidate application artifact is invalid for upgrade purpose.
+   * @throws NotFoundException if any object related to upgrade is not found like application/artifact.
+   * @throws Exception if there was an exception during the upgrade of application. This exception will often wrap
+   *                   the actual exception
+   */
+  public ApplicationId upgradeLatestApplication(ApplicationReference appRef,
```

Comment:
can you add a `@deprecated` annotate to the other upgrade method? We don't want any new code to call it.